### ISSUE TITLE
Chef-zero startup performances compared to chef-solo are awful

### ISSUE THREAD

**Issue by gionn** [chef/chef#3803]
It's nice that you are trying to get people move from chef-solo to chef-zero/local-mode, but with the actual performances I don't think we are going anywhere.

```
[15:49:01.662613 ] [ INFO]: Started chef-zero at chefzero://localhost:8889 with repository at /var/chef/cache
[15:49:01.663691 ] [ INFO]: Forking chef instance to converge...
[15:49:01.689879 ] [ INFO]: *** Chef 12.4.1 ***
[15:49:01.690425 ] [ INFO]: Chef-client pid: 5410
[15:49:03.381683 ] [ INFO]: HTTP Request Returned 404 Not Found: Object not found: chefzero://localhost:8889/nodes/stag-1914102-nodejs-mongodb-helloworld.appstaging.cloudesire.com
[15:49:03.384887 ] [ INFO]: Setting the run_list to ["cd-base", "cd-appdeploy", "cd-mongodb", "cd-nodejs"] from CLI options
[15:49:03.391446 ] [ INFO]: Run List is [recipe[cd-base], recipe[cd-appdeploy], recipe[cd-mongodb], recipe[cd-nodejs]]
[15:49:03.391846 ] [ INFO]: Run List expands to [cd-base, cd-appdeploy, cd-mongodb, cd-nodejs]
[15:49:03.392133 ] [ INFO]: Starting Chef Run for stag-1914102-nodejs-mongodb-helloworld.appstaging.cloudesire.com
[15:49:03.392505 ] [ INFO]: Running start handlers
[15:49:03.392714 ] [ INFO]: Start handlers complete.
[15:49:03.393948 ] [ INFO]: HTTP Request Returned 404 Not Found: Object not found:
[15:50:34.192062 ] [ INFO]: Loading cookbooks [cd-base@0.4.0, cd-appdeploy@0.3.0, cd-mongodb@1.0.0, cd-nodejs@0.2.0, apt@2.7.0, build-essential@2.2.3, hostname@0.3.0, maven@1.3.0, lvm@1.3.7, locale-gen@0.0.3, swap@0.3.8, sudo@2.7.2, sysctl@0.6.2, openssh@1.4.0, java@1.32.0, postfix@3.7.0, cd-nginx@0.0.1, flyway-cli@0.3.1, database@2.1.10, mongodb@0.16.2, git@4.0.2, hostsfile@2.4.5, ark@0.9.0, windows@1.36.6, ohai@2.0.1, iptables@1.0.0, nginx@2.7.6, ubuntu@1.1.8, mysql@5.6.3, postgresql@3.4.14, aws@2.7.2, xfs@1.1.0, mysql-chef_gem@0.0.5, yum@3.6.3, python@1.4.6, runit@1.7.2, dmg@2.2.2, yum-epel@0.6.2, 7-zip@1.0.2, chef_handler@1.2.0, bluepill@2.3.1, yum-mysql-community@0.1.17, openssl@4.3.2, packagecloud@0.0.19, rsyslog@2.1.0, chef-sugar@3.1.1]
[15:50:37.778632 ] [ INFO]: Storing updated cookbooks/cd-base/recipes/debian.rb in the cache.
[15:50:37.878875 ] [ INFO]: Storing updated cookbooks/cd-base/recipes/ssh.rb in the cache.
[15:50:37.977315 ] [ INFO]: Storing updated cookbooks/cd-base/recipes/ssl.rb in the cache.
[15:50:38.278576 ] [ INFO]: Storing updated cookbooks/cd-base/recipes/firewall.rb in the cache.
[15:50:38.377963 ] [ INFO]: Storing updated cookbooks/cd-base/recipes/logger.rb in the cache.
[15:50:39.079003 ] [ INFO]: Storing updated cookbooks/cd-base/recipes/dns.rb in the cache.
[15:50:39.179879 ] [ INFO]: Storing updated cookbooks/cd-base/recipes/agent_windows.rb in the cache.
[15:50:39.279209 ] [ INFO]: Storing updated cookbooks/cd-base/recipes/users.rb in the cache.
...
[15:54:35.232144 ] [ INFO]: Processing bash[apt-get-update at compile time] action run (apt::default line 39)
```

over 4 minutes before actually doing something useful on the node. :sleepy:

and then with chef-solo:
```
[16:00:06.541206 ] [ INFO]: Forking chef instance to converge...
[16:00:06.567303 ] [ INFO]: *** Chef 12.4.1 ***
[16:00:06.568125 ] [ INFO]: Chef-client pid: 6892
[16:00:08.411502 ] [ INFO]: Setting the run_list to ["cd-base", "cd-appdeploy", "cd-mongodb", "cd-nodejs"] from CLI options
[16:00:08.425035 ] [ INFO]: Run List is [recipe[cd-base], recipe[cd-appdeploy], recipe[cd-mongodb], recipe[cd-nodejs]]
[16:00:08.425455 ] [ INFO]: Run List expands to [cd-base, cd-appdeploy, cd-mongodb, cd-nodejs]
[16:00:08.425739 ] [ INFO]: Starting Chef Run for stag-1914102-nodejs-mongodb-helloworld.appstaging.cloudesire.com
[16:00:08.426330 ] [ INFO]: Running start handlers
[16:00:08.426559 ] [ INFO]: Start handlers complete.
[16:00:09.890759 ] [ INFO]: Processing bash[apt-get-update at compile time] action run (apt::default line 39)
```

3 seconds. :dancer:

**Comment by lamont-granquist** [comment]
So if it did that every time we probably would have fixed it by now since it would have aggravated people who use it all the time.  We most likely at least need debug output to see what its doing, and it looks like something might be hitting a 30 second timeout of some sort.  Also the O/S that you're on would help, and if any virtualization is involved anywhere.

**Comment by gionn** [comment]
I can further investigate enabling debug output for the 1 minute and half between:

```
[15:49:03.392714 ] [ INFO]: Start handlers complete.
[15:50:34.192062 ] [ INFO]: Loading cookbooks [...]
```

but the filling of the chef-client cache is really a waste of time, since the cookbooks files are already somewhere on the current node (4 minutes for that), and I know this is due to the chef-client have no idea that the chef-server is actually a chef-zero instance running on the same machine.

I've also tried to unpack my cookbooks.tar.gz directly on the cookbooks cache folder as someone suggested somewhere, but it caused an incomprehensible catastrophe of errors at the first run, solved only by wiping cookbooks cache entirely (if you think this is interesting enough, I can re-trigger and open a new issue).

Thanks.

**Comment by gionn** [comment]
```
[2015-08-21T08:53:51+00:00] DEBUG: Synchronizing cookbooks
[2015-08-21T08:53:51+00:00] DEBUG: Chef::HTTP calling Chef::HTTP::JSONInput#handle_request
[2015-08-21T08:53:51+00:00] DEBUG: Chef::HTTP calling Chef::HTTP::JSONToModelOutput#handle_request
[2015-08-21T08:53:51+00:00] DEBUG: Chef::HTTP calling Chef::HTTP::CookieManager#handle_request
[2015-08-21T08:53:51+00:00] DEBUG: Chef::HTTP calling Chef::HTTP::Decompressor#handle_request
[2015-08-21T08:53:51+00:00] DEBUG: Chef::HTTP calling Chef::HTTP::Authenticator#handle_request
[2015-08-21T08:53:51+00:00] DEBUG: Chef::HTTP calling Chef::HTTP::RemoteRequestID#handle_request
[2015-08-21T08:53:51+00:00] DEBUG: Chef::HTTP calling Chef::HTTP::ValidateContentLength#handle_request
[2015-08-21T08:53:51+00:00] DEBUG: #<ChefZero::RestRequest:0x000000051d4ed0 @env={"SCRIPT_NAME"=>"", "SERVER_NAME"=>"localhost", "REQUEST_METHOD"=>"POST", "PATH_INFO"=>"/envi
ronments/cloudesire/cookbook_versions", "QUERY_STRING"=>nil, "SERVER_PORT"=>8889, "HTTP_HOST"=>"localhost:8889", "rack.url_scheme"=>"chefzero", "rack.input"=>#<StringIO:0x000000051d5010>}, @rest_base_prefix=["organizations", "chef"]>
[2015-08-21T08:53:51+00:00] DEBUG: {"run_list":["cd-base","cd-appdeploy","cd-mongodb","cd-nodejs"]}
[2015-08-21T08:54:37+00:00] DEBUG: 
--- RESPONSE (200) ---
{
  "cd-base": {
    "recipes": [
      {
        "name": "users.rb",
        "path": "recipes/users.rb",
        "checksum": "cb3b55640aa4cd411930f495adbbd0cd",
        "specificity": "default",
        "url": "chefzero://localhost:8889/file_store/repo/cookbooks/cd-base/recipes/users.rb"
      },
```

It just needs some time to generate the response, we are plenty of cookbooks and the instance is the equivalent of a t2.small (cpu is stuck at 99% all the time).

**Comment by gionn** [comment]
Sorry, I forgot to mention that we are on a private OpenStack, KVM-based, Ubuntu 14.04.

**Comment by lamont-granquist** [comment]
Is the 99% CPU mostly userspace or kernelspace?  If its kernelspace then you're likely getting crappy I/O perf from virtualization and should probably look at making sure you're using paravirt_ops (and i believe you want HVM-with-PV-drivers these days rather than running fully paravirtualized).

**Comment by gionn** [comment]
Mostly user-space, near zero I/O wait all the time (plenty of memory for kernel to cache).

I want to reiterate what stated in the first post: with *exactly* the same configuration between chef-solo and chef-client with local-mode, the first perform so far *more* better than the second, logs demonstrate it.

I have a `solo.rb` in my `/etc/chef`:

```
recipe_url '$$$_COOKBOOKS_$$$'
environment 'cloudesire'
environment_path '/var/chef/environments'
cache_path '/var/chef/cache'
# See https://github.com/chef/chef/issues/3802
cookbook_path '/var/chef/cache/cookbooks'
json_attribs '/etc/chef/first-boot.json'
log_level :info
```

I am running chef-solo with `chef-solo`,
and chef-zero with `chef-client --local-mode -c /etc/chef/solo.rb`.

If this situation is not improved, we are stuck on relaying on chef-solo (we have strict delivery times when deploying a new application stack on a vanilla ubuntu VM, under 10 minutes).

Thanks for your help.

**Comment by lamont-granquist** [comment]
Somewhere there's an open issue that local-mode shouldn't copy the cookbooks up to the 'server' only  to place them in exactly the same filesystem -- at the very least its known and has been discussed.

It looks like you have a pile of time where its just 'downloading' the cookbook files, though, which which is a perf issue that I don't think we're aware of.  The fact that its user space is also weird.
It might be that we're doing SSL over loopback which is a little pointless, but that'd mean you'd have a fairly rediculously low amount of CPU available (on an old t1.micro image I wouldn't be surprised).

**Comment by ranjib** [comment]
chef-zero has always been pretty slow for more than a handful of cookbooks. It will be nice to have local mode where cookbook management logic is similar to solo, but the search functionality is based on chef-zero

**Comment by gionn** [comment]
In fact, we started using chef-zero under test-kitchen without even noticing it, but we stepped back when trying it into production.

**Comment by lamont-granquist** [comment]
So, complaining about it is less useful than letting us know where the performance issue actually is, or giving us a good replication case.  Right now I can only speculate about which code to patch, and my guess may very well be incorrect.  I use it entirely under TK so it just works fine for me.

**Comment by eric-tucker** [comment]
I'd like to add that I agree it's a huge waste of time for Chef zero to copy the cookbooks to the cache before running.  This seems like behaviour duplicated from Chef server (copying cookbooks from server to local cache).

**Comment by lamont-granquist** [comment]
This may be fixed in 12.5 according to @jkeiser

**Comment by lamont-granquist** [comment]
AFAIK should be fixed, probably cut a new ticket if 12.5.0 is still bad and/or just not good enough.

**Comment by gionn** [comment]
I will take a shoot at the new version, thanks to everyone!
